### ISSUE TITLE
Preparing an animation of ell tube evolution takes forever

### ISSUE THREAD

**Issue by pgagarinov** [SystemAnalysisDpt-CMC-MSU/ellipsoids#13]
Deadline - 1st of November, by that time both the changes and all the tests should be integrated into "master" branch
--------------------------------------------------------------------------------------
elltool.demo.test.control.anim6 seems to work very slowly, thanks to performance of EllTube.cut method  and some others. We need to do a major performance optimization. One needs to use a profiler (profile on; run calculation... ; profile viewer) to find the performance bottlenecks and fix them. It can be that some portions of the code need to be refactored for better performance. 

As part of this task one needs to write a few performance tests in +gras\+ellapx\+smartdb\+test\+mlunit\SuiteEllTube.m for the methods that will be optimized. The idea for writing the performance test can be as follows: we perform an animation of already calculated tube in two different ways: a) ugly but fast via extracting the ellipsoidal matrices from the tube and doing the plotting manually via low-level functions from gras.geom.* packages. b) via EllTube.cut. Then we compare the runtimes of these two methods and check that the difference is less than x% where x% is fixed in the test. The goal is to make the overhead x less than at least 25%.

Once the optimization is done we will uncomment a call to elltool.demo.test.run_demo_tests from

in +elltool\+demo\+test\run_tests.m. Run now this call is commented out so that the test pack for ellipsoids takes under 2 hours to run.

**Comment by akskir** [comment]
Good evening. I have a few questions:
1. I don't really understand what method ReachContinuous.get_goodcurves() do.
2. As I understand we made a new tubes for all times in gcTimeVec  and for all shifts of x0. Also in cycle we push in videoFrame the end of the tube, thats how we make a video?
3. What I need to fix: this example(anim6) or function EllTube.cut? I ask this because EllTube.cut doesn't looks like the main problem in this code. Yes, it works not so fast, but building of new ReachContinuous object really have to work much longer.
4. Also maybe I need to optimize EllTube.cut method with just 1 argument?
Thanks for answering.

**Comment by akskir** [comment]
Also as I understand ReachContinuous.get_goodcurves()  returns something like the center line of the tube, but i'm not sure. Cause we use use it like a shift after.

**Comment by akskir** [comment]
Now it mostly doesn't matter, because I'm going deeper in Ellipsoidal Toolbox, and how I understand the main problem is in the createInstance function wich we use in interp, wich was used in cut function.

**Comment by pgagarinov** [comment]
1) Added a help header for this method along with minor cosmetic changes to anim6 in "master" branch - please do reabase. Please let me know if this doesn't help.

2) 

a) As the first step we build an approximation of reachability tube. on [t_0,t_0+T] with a set of ellipsoidal tubes, one for each direction l_i from dirsMat. As a result of this approximation we have a list of touch curves in R^{n}, one curve for each l_i. We take a touch x(t) curve corresponding to l_1. 

b) Then we start shifting a time span [t_0,t_0+T] forward bit by bit by certain delta and for each shifted position we 
    - build an approximation of reachability domain with x(t_0+delta) +B_1(0) as initial set. 
   -  cut ellipsoidal tubes so that they all end at t_0+T+delta. 
   -  plot an external approximation of reachability tube.


First of all, I'm not sure we need "cut" call here at all as the tubes should already end at t_0+T+delta. 

There is an old version of "anim6" in products\elltoolboxcore\control\test\anim6 from ET v.2 (https://github.com/SystemAnalysisDpt-CMC-MSU/ellipsoids/archive/v2.0.zip). As a reference you can grab this old anim6 and run it against the very first version of the toolbox that works much faster (but contains bugs and much less functionality) (https://github.com/SystemAnalysisDpt-CMC-MSU/ellipsoids/releases/download/v1.1.3/elltbx_1_1_3_full.zip)

So please investigate if we can drop "cut" and have the same result. There can be that a student who translated anim6 from old toolbox to the new toolbox API made a mistake. If there is a mistake - this is for you to find.

Secondly, I'm not sure we need to build an approximation of the original set for all l_i if we only use the first tube corresponding to l_1 anyway.

3) You need to make sure that a) anim6 was translated from the old anim6 without any mistakes (though I do not suspect that there are any). b) see how we can optimize the new anim6. c) if any bugs in cut or other methods are found (performance or not-performance related) - they need to be covered with tests and fixed.

4) What do you mean?

5) 
> I understand ReachContinuous.get_goodcurves() returns something like the center line of the tube. 
No, it returns a list of touching lines, not center lines.

6) 

>how I understand the main problem is in the createInstance function wich we use in interp, wich was used in cut function.

So what is percentage of total runtime has to do with "cut" method?

**Comment by akskir** [comment]
1) As I understand everything was translated correctly
2) We really don't need so many directions, because we use just one. When I changed nDirs to 1, time of this example become to be ~23 minutes
3) Also, if we really start to use just one direction, I can change varargin{1} from number of directions to angle of direction or we can generate direction by random
4) How I understand we need to use cut function, because without it we make a video about change full tube, not only reach set
5)Percentage of total runtime "cut" method is very small, about 4%. Full time of "cut" method is about 59.514 seconds, when all time is 1399 seconds. In profiler I see, than the main problem is in the AReach.MakeEllTubeRel, wich call several another functions and in the end too often uses ode45.

**Comment by akskir** [comment]
If you want, I can send to you report of profiler work in pdf.

**Comment by pgagarinov** [comment]
2) We use just one from the first call but then

    rsObj = elltool.reach.ReachContinuous(sys, x0EllObj, dirsMat,...
        [startTime endTime],'isRegEnabled',true, 'isJustCheck', false ,'regTol',1e-3);

uses all the directions, isn't that so?

5) Where is the performance bottleneck then? What is the main contributors to the total calculation time? ode45? 

I can generate the report by myself but I would rather have your opinion on whether we have any room for optimization?

**Comment by akskir** [comment]
Yes, more directions give us better approximation of tube, but, how I think, we don't need if on this example, because we use only one curve line in cycle.
The main problem is ode45, but if look only on self time most long-working functions is parseparext and parseparams, but it's not more then 10% of all time.

**Comment by pgagarinov** [comment]
Also, can you please have a look at anim1 - there we do not have that many calls to ReachContinuous and I suspect that cut performance is a bottleneck there. If so - is there any room for performance optimization in cut?

**Comment by akskir** [comment]
Ode45 works fast, but we call it 400 times. Ok, I'll see and try to optimize it

**Comment by pgagarinov** [comment]
Ok, let us forget about anim6 and concentrate on anim1 where cut is a bottleneck, not ode45.

**Comment by akskir** [comment]
Looking at profiler report, I can say, that on anim1 main bottleneck is "plotByEa", "cut" method use only 16% of all time, but we call it about 100 times. Also I have already tried to optimize "cut" and how I understand "EllTube.interp" is longest-running call in "cut".

**Comment by pgagarinov** [comment]
How much plotByEa does take in percentage terms?
The final goal is reduce the overall runtime of anim1 so if plotByEa is a bottle neck then we should concentrate on it first.

**Comment by akskir** [comment]
"plotByEa" takes little bit more than a half of a time

**Comment by pgagarinov** [comment]
As you understand we re-draw some part of the time over and over instead of adding vertices/edges to the existing patch object. The latter should be significantly faster. As I said in the description let us try to implement anim1opt
that 
1) Extracts ellipsoidal matrices from ellipsoidal tube via getData
2) Plots them directly using the optimal way (as described above - via adding points to already displayed patch instead of re-drawing the whole tube over and over again)
3) Compare runtimes of anim1 and anim1opt

Once you make sure that anim1opt runs significantly faster - I will consider this task to be done. After all this is a warm-up, not a full scale task and you need time for the full scale task.

-------------------------------------
So the goal of this task is to have anim1opt that does the same as anim1 but much faster and in ugly way (steps 1-3), just a proof of concept. Of course "ugliness" doesn't mean that the code should be ugly, it should comply with all the coding conventions. "Ugliness" means that you do not have to write tests and use OOP of functional programming concepts, you can do everything directly in anim1opt but result should be the same as anim1.

**Comment by akskir** [comment]
Ok, shall I make only new getAnimation function? Because on anim1 we only make a system, ReachContinuous object and call getAnimation function.

**Comment by akskir** [comment]
And I need to ask what ReachContinuous.refine do? It change direction?

**Comment by pgagarinov** [comment]
Yes, new getAnimation function

Refine does have a help header - "doc elltool.reach.ReachContinuous.refine"

**Comment by pgagarinov** [comment]
Please
1) use a separate variable for numStartPloting:numEndPloting
2) iTimeSteps -> iTimeStep
3) use constants in upper case with underscores 
4) at the begining of getAnimation for all hardcoded values like "1000", "[39.6 -35.6]" or "0.3"
5) use smart indentation - select all code with ctrl+A and then press ctrl+I.
6) all lines should be under 76 symbols - use "..."
7) do not use round brackets in "if" conditions
8) QArray -> bigQArr or qBigArr

**Comment by akskir** [comment]
I've pushed new version of anim1opt, but I don't use transparency, because if plot full object and change it [transparency] then it start to work much longer or it would be necessary to use very low level matlab function than rely on openGl, but how I think it is too complicated for this example... Please check is everything OK

**Comment by akskir** [comment]
Now it works more then 5x time faster

**Comment by pgagarinov** [comment]
10) MLint (Matlab code checker gives a few warnings - on one of seminars I asked to make sure that there are no warnings before committing something. There should be a green square in the right upper corner of editor window.

11) When I run anim1opt I get this error:

Warning: No video frames were written to this file. The file may be invalid. 
> In VideoWriter/close (line 267)
  In VideoWriter/delete (line 202)
  In elltool.demo.test.control.anim1opt>getAnimation (line 55)
  In elltool.demo.test.control.anim1opt (line 38) 
Undefined function or variable 'approxType'.

Error in elltool.demo.test.control.anim1opt>getAnimation (line 65)
graphObjectName =  ['Reach Tube: by ', char(approxType(1))];

Error in elltool.demo.test.control.anim1opt (line 38)
writerObj = getAnimation(firstRsObj,writerObj,[0,5],axisConfVec);

12) Hm, this is very strange that transparency change takes that long. 
Did you run "opengl hardware" before changing transparency?

Also, have you actually tried to construct a relation data plotter object and then passing it into plotByEa like this:

plObj=smartdb.disp.RelationDataPlotter()
reachObj.plotByEa(plObj);

%change transparency in a loop using a patch object handle from plObj

?

**Comment by akskir** [comment]
Wow, this was my fault, I have corrected it, and now I'll try to this task via changing transparency value,  but I don't really understand how to do it, because I send to patch function Faces and Vertex matrix so I need to work with them to find vertices, where I need to change transparency to zero? Is it true? And what is "opengl hardware", how I can work with it?

**Comment by akskir** [comment]
When I say that it works to long, I mean that I tried to use PlotPropProcessor class, but it was not I expect to. And how I understand if I want to use this class I need to redraw all tube. Please correct me if I'm wrong.

**Comment by pgagarinov** [comment]
I see but PlotPropProcessor doesn't have anything to do with what I suggest. RelationDataPlotter is just a container for handles of all graphical objects produced by plotByEa/Ia, plotEa,plotIa methods.

RelationDataPlotter object provides "getPlotStructure" method that returns a structure with 3 fields:

figHMap   - figure handles
figToAxesToHMap - axes handles for each figure
figToAxesToPlotHMap - graphical object handles for each axes for each figure

You need to use figToAxesToPlotHMap  field to find all patch objects that correspond to all kinds of tubes. Please note that touch lines are also plotted via patch object (zero-area facets with just edges). You should use "Type" field of a graphical object:

>> h=plot(1:10);get(h,'Type')

ans =

line

For tubes this field will contain "Patch".

Redrawing of the tube is not required, you just change FaceVertexAlphaData for a patch object to hide certain faces. Type doc patch for more details.

opengl is Matlab built-in command, type "doc opengl" for more details.

**Comment by akskir** [comment]
So I need to rewrite this example?

**Comment by pgagarinov** [comment]
I would rather suggest to implement anim1opt2 via transparency, this shouldn't take long give a simplicity of the task. 

Your current idea is to implement the whole new method because transparency didn't work. But this is not practical because you didn't try transparency in a right way (without re-creating the tube) and a potential implementation is going to be much simpler - just call the same method and just make a part of the tube invisible. I can't be simpler that this. To be honest I didn't understand your explanation about transparency requiring low-level opengl calls. I though you would just make certain edges 100% transparent.

Of course if you prove the transparency is much slower than yours - then we consider the task completed.

**Comment by akskir** [comment]
I've pushed new version that works with patch object and it is really faster. Please check and say to me is everything ok?

**Comment by akskir** [comment]
It is not much faster then my method, but it works ~10 seconds faster.

**Comment by pgagarinov** [comment]
Please commit both versions - anim1opt1 and anim1opt2.

In transparency-based please make sure that you 

13) do not use hard-coded indices to access stuff like this

secondMap = firstMap(firstKey{1});
secondKey = secondMap.keys;
plotGroup = secondMap(secondKey{1});

and use an adaptive code that would work even if there are more than 1 elements there. Or throw an exception if there is more than 1 element. 

14) plotGroup variable doesn't seem to be  a scalar - please change its name according to our naming policy (see Wiki). Same goes for typesGroupStruct, secondKey, numElementsInGroup->nElements, sizeVert ->nVertices

15) Lines should be under 76 symbols.

**Comment by akskir** [comment]
Please check is everything ok now

**Comment by pgagarinov** [comment]
16) Using while to check if condition is met is very wierd, please use if. Also, in your case if condition is not met
nothing happens but you need to throw an exception using modgen.common.throwerror;

+while (numel(mapObj(SkeyData{1})) == 1) &&...
-secondKey = secondMap.keys;		+        (ishandle(mapObj(SkeyData{1})) == 0) 
-plotGroup = secondMap(secondKey{1});		+    mapObj = mapObj(SkeyData{1});
-set(plotGroup, 'Visible', 'off')		+    SkeyData = mapObj.keys;
+end

17) Array suffix is only for arrays of dimension >=3. Please study the variable naming convention and use the proper suffixes 
(they are different for numeric and cell variables)

18) STypesGroupDataArray ,SkeyData: only strutures' names and constants can start with a capital letter

**Comment by akskir** [comment]
I made throwerror in anim1opt2 and fix small fault in anim1opt1, also chenged names of variables

**Comment by pgagarinov** [comment]
What are types are sizes of these variables?

**Comment by akskir** [comment]
In this example: keyC is cell 1x1, plotGroupArray is Group 1x8. I understand why not -Array. But keyC can be cell 1xn where n is number of objects in map.

**Comment by pgagarinov** [comment]
Yes, the names you suggested are just fine.

A few additional problems to fix:

19) If you use some packaged function/class more than once - use import.
20) All lines should be under 76 symbols.
21) Please fix the problems found by mlint (Matlab Code Checker) in anim1opt1.


And finally, why anim1opt1 works so fast comparing to anim1?

**Comment by akskir** [comment]
It works faster because we don't redraw  full tube. This is the main reason, also we don't need to get EllTube and EllTubeProj every time, but do it just once for every RC object. Also I can say that because of the really big steck of calls, that we have on normal plotting by plotByEa.
I have question: Can I do a rebase now? I mean that is master now in good condition after some mistakes that was made by my colleagues?

**Comment by pgagarinov** [comment]
Yes, you can, master is clean now.

**Comment by akskir** [comment]
I received Jenkins logs and it says that built built "Fixed" and in build.log I have  BUILD SUCCESSFUL, is this mean that everything ok? (I ask this because I don't want to break master.) Also is everything ok in last commit?

**Comment by pgagarinov** [comment]
You branch doesn't look like it is rebased to the latest version of master. I assume that is because prior to rebase your forgot to do fetch. PLease note - unless you do fetch remote branches (in remove/origin) including are not updated. So I guess you rebased to some old version of master that was in your local repo from old times. 

How I know? 

Last commit in master is 

SHA-1: 94391bc7d30954701e7b338be7d2744eaa9595f7

* Enhancement: fixed a potential problem in auxdfevalpcomp.m with file write lock when jobStartup.m file is created even when no parallel processes is created
made on 13/11/2015

while the last commit that precedes your commits is
SHA-1: 1d6680c192ac79f929012666246de7994ed0e7b7

* issue #13
Enhancement: a few cosmetic changes
made on 29/10/2015

You can always login to github using webrowser and see the check the latest commit for mastet branch just to be 100%.

**Comment by akskir** [comment]
What shall I do after rebase?

**Comment by pgagarinov** [comment]
You should proceed according to the procedure which says that you need to wait for the test results.

**Comment by pgagarinov** [comment]
Why do you think that?

**Comment by akskir** [comment]
Because my commit, that was made 14 days ago, is upper that commit that was made by you 6 days ago... Nonetheless, all tests was successful, and as I understand was made successful too, so what shall I do? Try to make rebase and if everything up-to-date do merge?

**Comment by pgagarinov** [comment]
Can you tell me hashes of those commits?

**Comment by akskir** [comment]
SHA-1 of my commit: bac1c71238c88aff66baf031917d905900c1ec6f
SHA-1 of your commit: 94391bc7d30954701e7b338be7d2744eaa9595f7

**Comment by pgagarinov** [comment]
Right, but this is exactly how rebase works - it attaches your commits to the tip of master branch. Of course the attached commits are not exactly the same as the original ones. It is merge operation that preserves an order of commits. 


Nikita Zuyanov has just merged into master, let us wait for the tests in master to pass and then you need to rebase again, sorry, Nikita was faster this time.

**Comment by akskir** [comment]
Understand.Ok, now I need to wait for Nikita's branch tests pass, after do rebase wait for my tests pass and make a merge

**Comment by pgagarinov** [comment]
Yes

**Comment by pgagarinov** [comment]
Master is ready, you can rebase now.

**Comment by pgagarinov** [comment]
You merged into master incorrectly so I had to revert your changes. You made two mistakes.

1) Incorrect merge of your branch into master
2) Push of incorrectly modified master into master. 

You may do something incorrectly but you always need to check you graph structure prior to doing push. You graph structure of master after merge should look like a single "petlya" - see how Ilya and Nikita did their merges. It is obvious that graph of your changes in master doesn't look right - just by visual comparison. Next time please double-check prior to doing merge.


Now, please merge again, strictly following the procedure and do not push until you have a single "petlya". 

Thanks.

**Comment by pgagarinov** [comment]
This is how I should look like in master after a correct merge.

![1](https://cloud.githubusercontent.com/assets/4868370/11279545/4f6ecd88-8f02-11e5-9d16-b2e65e4419dc.png)

**Comment by akskir** [comment]
![2015-11-19 1](https://cloud.githubusercontent.com/assets/15004295/11280282/5c136306-8f06-11e5-8b67-1e948d4209b7.png)
Now everything must be ok

**Comment by pgagarinov** [comment]
Yes, looks fine.

**Comment by akskir** [comment]
Now all tests passed, is this issue closed?

**Comment by pgagarinov** [comment]
Yes, thanks.
